feat: AWS OIDC integration role `policy-preset` for identity center integration

flyinghermit opened the pull request:
Adds a new `policy-preset` flag in `teleport integration configure awsoidc-idp` command. 
`policy-preset` flag is optional and policy assignment will be ignored if the flag is not used, if a flag is defined with an empty string value or an `unspecified` value.

As a first supported preset type, this PR adds a preset for AWS Identity Center integration which can be configured as `--policy-preset=aws-identity-center`.

Feature implemented as a part of identity center integration https://github.com/gravitational/teleport.e/issues/4839

changelog: Added a new `policy-preset` flag in the `teleport integration configure awsoidc-idp` command which can be used to configure a given AWS IAM role with permissions required for a supported Teleport and AWS integration.

flyinghermit committed:
feat: adds a new 'policy-preset' flag to 'integration configure awsoidc-idp' command

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/aws-oidc-preset-policy (pending feature parity)

flyinghermit committed:
adapt to new awsactions

flyinghermit committed:
output formatting and golden test

flyinghermit committed:
run ui prettier, fix getAwsOidcConfigureIdpScriptUrl test

flyinghermit committed:
comments cleanup, mend TestBuildAWSOIDCIdPConfigureScript test

GavinFrazar commented on lib/integrations/awsoidc/idp_iam_config_test.go:
```diff
@@ -354,6 +476,24 @@ func (m *mockIdPIAMConfigClient) CreateRole(ctx context.Context, params *iam.Cre
+func (m *mockIdPIAMConfigClient) PutRolePolicy(ctx context.Context, params *iam.PutRolePolicyInput, optFns ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error) {
+	doesNotExistMessage := fmt.Sprintf("Role %q does not exist.", *params.RoleName)
+	if _, ok := m.existingRoles[aws.ToString(params.RoleName)]; !ok {
+		return nil, &iamtypes.EntityAlreadyExistsException{
```
```suggestion
		return nil, &iamtypes.NoSuchEntityException
```

GavinFrazar commented on web/packages/teleport/src/config.ts:
```diff
@@ -1245,6 +1245,7 @@ export interface UrlAwsOidcConfigureIdp {
   roleName: string;
   s3Bucket?: string;
   s3Prefix?: string;
+  policyPreset?: string;
```
should we use the enum type instead of string here?

GavinFrazar commented on lib/integrations/awsoidc/idp_iam_config.go:
```diff
@@ -113,6 +144,14 @@ func (r *IdPIAMConfigureRequest) CheckAndSetDefaults() error {
 	r.ownershipTags = tags.DefaultResourceCreationTags(r.Cluster, r.IntegrationName)
 
+	switch r.IntegrationPolicyPreset {
+	case "":
+		r.IntegrationPolicyPreset = PolicyPresetUnspecified
```
It's not clear to me why we need the `"unspecified"` instead of just using `""`?

flyinghermit replied:
In this particular case, handling both empty string `""` and spelled out default `unspecified` value is not that important, I went ahead and removed it in https://github.com/gravitational/teleport/pull/47164/commits/2ee21fea32f04a1b38f16cacbe626a2906855297

To answer your question on when both empty string `""` and `unspecified` would be helpful - `""` and `unspecified` lets us distinguish set and unset value. Say when `tctl` ships with some constsants/enums in version `v1` and we support that feature in the web ui in `v2`. And since proxy and auth can have different versions, it can be tricky in auth to identify if web ui deliberately sent an empty string or is in a version that does not support the feature altogether.

GavinFrazar commented on web/packages/teleport/src/services/integrations/types.ts:
```diff
@@ -373,6 +373,15 @@ export type AwsOidcDeployServiceRequest = {
+ * to the AWS IAM role created for the OIDC integration.
+ */
+export enum AwsOidcPolicyPreset {
+  Unspecified = 'unspecified',
```
Does it make more sense to use the empty string for the `Unspecified` variant?

```suggestion
  Unspecified = '',
```

GavinFrazar commented on lib/cloud/aws/policy_statements.go:
```diff
@@ -433,3 +433,48 @@ func StatementAccessGraphAWSSync() *Statement {
+
+// StatementForAWSIdentityCenterAccess returns AWS IAM policy statement that grants
+// permissions required for Teleport identity center client.
+// TODO(sshah): make the roles more granular by restricting resources scoped to
+// partivular AWS identity center region+arn.
```
I think to properly scope these permissions will require very careful conditional permissions, because it grants `iam:CreateRole` and `iam:AttachRolePolicy` which can be used to elevate permissions. Also note that any teleport discovery service will be able to get a token to assume the IdP integration role, so it may be time to think about scoping the trust policy further.

flyinghermit replied:
Definitely. This is a hard TODO I set and I will resolve it before release. Trent is coming back from vacation next week and I plan to sync with him to figure out the final permissions (the provisioning service is mostly done but we have some more rough edges to scope down) and conditions we can apply here. 

The gist is: most of the actions will be scoped to the identity center region + ARN we ask from user. The role create and attach policies are risky but required since we provision AWS role on demand when access request is granted in Teleport or when user is enlisted to an access list. During test, I found there are some actions that require organization-wide resource scope so I want to properly take time and plan for them, hence it is currently loosely defined to address all the permissions we know we require for the identity center integration.

Happy to add some minimum constraints in this PR if you think that is still important before we merge to master.

GavinFrazar replied:
ah ok, I'm fine with the permissions for now since it doesn't do anything yet.

I am more concerned about the existing integration trust policy which allows for any teleport discovery service to assume the role though. I don't see mention of that in the RFD, so i'll take this comment to the RFD

tcsc replied:
Is there some way we can prevent this from actually being exercised until we address the permission scope? Or at least force it to issue a warning if it gets hit? 

I'm just concerned in case we fill in the rest of the Identity Center work and forget to  revisit this.

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/aws-oidc-preset-policy

flyinghermit committed:
address review:
- use AwsOidcPolicyPreset in UrlAwsOidcConfigureIdp
- use NoSuchEntityException instead of EntityAlreadyExistsException in PutRolePolicy mock
- fix typo

flyinghermit committed:
use empty string as unspecified value

marcoandredinis reviewed (approved):
I was wondering if, with the current design, easily evolve to have multiple Policies being added to the IAM Role.
I would say it's pretty easy to change, right?
(not asking for it to be done done, just wondering if we are not making it difficult to do it in the future)

For example, we could present a single command to the user when setting up multiple features.
Eg a new flow where you add the Integration and configure Identity Center, App Access and EC2 Auto Discover.

flyinghermit commented on lib/cloud/aws/policy_statements.go:
```diff
@@ -433,3 +433,48 @@ func StatementAccessGraphAWSSync() *Statement {
+			"sso:DescribeAccountAssignmentCreationStatus",
+			"sso:DeleteAccountAssignment",
+			"sso:DescribeAccountAssignmentDeletionStatus",
+			"iam:AttachRolePolicy",
```
Definitely. See previous discussion here https://github.com/gravitational/teleport/pull/47164#discussion_r1788052960, I will resolve the permission modal at once.

flyinghermit commented on lib/integrations/awsoidc/idp_iam_config.go:
```diff
@@ -234,3 +283,24 @@ func createIdPIAMRoleAction(clt IdPIAMConfigureClient, req IdPIAMConfigureReques
+
+	switch req.IntegrationPolicyPreset {
+	case PolicyPresetAWSIdentityCenter:
+		policyName = "TeleportAWSIdentityCenterIntegration"
```
Why do you think it is a bad idea though?
It helps visually distinguish in the AWS console without looking at tags and created by and I was even thinking to eventually do that in the other existing policies as well.

flyinghermit commented on lib/integrations/awsoidc/idp_iam_config.go:
```diff
@@ -83,6 +87,33 @@ type IdPIAMConfigureRequest struct {
+// ErrAWSOIDCInvalidPolicyPreset is issued if provided policy preset
+// value is not supported.
+var ErrAWSOIDCInvalidPolicyPreset = &trace.BadParameterError{
+	Message: "--preset-policy defines an unknown preset value",
```
It would be nice but it adds a burden for us to update the message every time a new preset will be added and easy to overlook. The message is clear enough for user to reference in docs.

flyinghermit commented on lib/cloud/provisioning/awsactions/assign_policy.go:
```diff
@@ -0,0 +1,102 @@
+		RunnerFn: func(ctx context.Context) error {
+			_, err = clt.PutRolePolicy(ctx, input)
+			if err != nil {
+				if trace.IsNotFound(awslib.ConvertIAMv2Error(err)) {
+					return trace.NotFound("role %q not found.", req.RoleName)
+				}
+				return trace.Wrap(err)
```
Just looked into AWS api docs on possible error returned here https://docs.aws.amazon.com/IAM/latest/APIReference/API_PutRolePolicy.html#API_PutRolePolicy_Errors. There are two errors that the client has direct influence over, `NoSuchEntity` and the `MalformedPolicyDocument`. `NoSuchEntity` is covered above and policy document is hardcoded in Teleport. 

Given that, I think we are fine not converting other errors here and returning as is as they are not directly actionable in Teleport.

flyinghermit commented:
> I was wondering if, with the current design, easily evolve to have multiple Policies being added to the IAM Role.
I would say it's pretty easy to change, right?
> For example, we could present a single command to the user when setting up multiple features.
Eg a new flow where you add the Integration and configure Identity Center, App Access and EC2 Auto Discover.

Yeah swapping out input type from `string` to `[]string` should be enough to support that behavior.

flyinghermit committed:
add optional field comment

flyinghermit commented on lib/integrations/awsoidc/idp_iam_config.go:
```diff
@@ -56,6 +56,13 @@ type IdPIAMConfigureRequest struct {
+	IntegrationRole string
+
+	// IntegrationPolicyPreset is the name of a policy preset to be applied to the IntegrationRole.
+	IntegrationPolicyPreset PolicyPreset
```
https://github.com/gravitational/teleport/pull/47164/commits/e48651fdab9bbf49a582c1b950f46d45adf6942a

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/aws-oidc-preset-policy

flyinghermit commented:
I definitely broke my latest sync with master. Fixing that now!

flyinghermit committed:
Revert "Merge branch 'master' of github.com:gravitational/teleport into sshah/aws-oidc-preset-policy"

This reverts commit ad2a0904e9076f39e468083b906f101b1171283b.

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/aws-oidc-preset-policy

r0mant reviewed (approved):
Bot.
